src/main/java/net/dries007/tfc/api/capabilities/forge/ForgingHandler.java

```diff
@@ -0,0 +1,132 @@
+ */
+
+package net.dries007.tfc.api.capabilities.forge;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.Direction;
+import net.minecraft.util.ResourceLocation;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.util.LazyOptional;
+
+public class ForgingHandler implements IForging
+{
+    private final LazyOptional<IForging> capability = LazyOptional.of(() -> this);
+
+    protected ItemStack container;
+
+    public ForgingHandler(ItemStack container)
+    {
+        this.container = container;
+    }
+
+    public ForgingHandler()
+    {
+        container = ItemStack.EMPTY;
+    }
+
+    public ItemStack getContainer()
+    {
+        return container;
+    }
+
+    @Override
+    public int getWork()
+    {
+        CompoundNBT tag = container.getTag();
+        if (tag != null && tag.contains("forging"))
+        {
+            return tag.getCompound("forging").getInt("work");
+        }
+        return 0;
+    }
+
+    @Override
+    public void setWork(int work)
+    {
+        getTag().putInt("work", work);
+    }
+
+    @Nullable
+    @Override
+    public ResourceLocation getRecipeName()
+    {
+        CompoundNBT tag = container.getTag();
+        if (tag != null && tag.contains("forging") && tag.getCompound("forging").contains("recipe"))
+        {
+            return new ResourceLocation(tag.getCompound("forging").getString("recipe"));
+        }
+        return null;
+    }
+
+    @Override
+    public void setRecipe(@Nullable ResourceLocation recipeName)
+    {
+        if (recipeName == null)
+        {
+            getTag().remove("recipe");
+        }
+        else
+        {
+            getTag().putString("recipe", recipeName.toString());
+        }
+    }
+
+    @Override
+    @Nonnull
+    public ForgeSteps getSteps()
+    {
+        CompoundNBT tag = container.getTag();
+        if (tag != null && tag.contains("forging"))
+        {
+            return ForgeSteps.get(tag.getCompound("forging").getCompound("steps"));
+        }
+        return ForgeSteps.empty();
+    }
+
+    @Override
+    public void addStep(@Nullable ForgeStep step)
+    {
+        getTag().put("steps", ForgeSteps.get(getTag().getCompound("steps")).serialize());
+    }
+
+    @Override
+    public void reset()
+    {
+        CompoundNBT tag = container.getTag();
+        if (tag != null)
+        {
+            tag.remove("forging");
+        }
+    }
+
+    @Nonnull
+    @Override
+    public <T> LazyOptional<T> getCapability(@Nonnull Capability<T> cap, @Nullable Direction side)
+    {
+        return CapabilityForging.CAPABILITY.orEmpty(cap, capability);
+    }
+
+    /**
+     * Initialize tag if needed, returns a tag with forging data
+     * Only call this when adding work / forge step
+     */
+    private CompoundNBT getTag()
```

Comment:
I've done that in `#reset`.

That function should only be called by setters (hence the javadoc i've added above), when you are certain you're adding a tag to the item. It's common code for all 3 since we need it initialized.

Getters are not using that function and instead check for the existence of the tag without creation.

Reply:
Yes but that's the exact point - now every potential modification to the tag has to check "is this tag actually empty of any useful data, and if so, call reset". For instance, just doing `cap.setWork(0)` is not sufficient.

The idea is that the caller code (the code working with this capability) should not have to consider when doing various set operations, what the state of the data in the tag is. If the right combination (for instance, setting the work then unsetting it, or setting the recipe and removing it) are called, the tag should be empty *without* requiring `reset` to be called. Making it so that this is the caller's responsibility is asking for a future bug to appear where `reset` is not getting called from some call path.